### PR TITLE
Re-index performance improvements

### LINKED ISSUE #1628: Indexer load queue doesn't backpressure
### Problem description
When the indexer does a full reindex of the registry it's supposed to crawl through and keep up to 1000 datasets in a buffer, but crawling and ingestion should proceed at roughly the right rate because of akka backpressure.

However recent improvements to the crawl speed in the database have revealed that this doesn't actually happen - if the database crawl goes faster than the elasticsearch ingestion (usual) then by the end of a full recrawl the indexer ends up with 10s of thousands of datasets backed up, which takes up a lot of memory.

We need to fix this so that the crawl only ever gets 1000 datasets or so in front of the ingestion.

### Screenshot / Design / File reference
Log:
```
[INFO] [08/23/2018 07:30:39.713] [indexer-akka.actor.default-dispatcher-13] [RegistryCrawler(akka://indexer)] Crawled 64000 datasets from registry
[INFO] [08/23/2018 07:30:42.079] [indexer-akka.actor.default-dispatcher-16] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:30:48.762] [indexer-akka.actor.default-dispatcher-27] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:30:55.056] [indexer-akka.actor.default-dispatcher-26] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:31:00.296] [indexer-akka.actor.default-dispatcher-29] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[WARN] [08/23/2018 07:31:05.950] [indexer-akka.actor.default-dispatcher-30] [akka.actor.ActorSystemImpl(indexer)] Failure when indexing ds-dga-f07db0a8-de52-48c4-a115-9d8a4eca379f: Some(BulkError(mapper_parsing_exception,failed to parse [spatial.geoJson],null,0,null))
[INFO] [08/23/2018 07:31:05.951] [indexer-akka.actor.default-dispatcher-30] [akka.actor.ActorSystemImpl(indexer)] Excluded dataset ds-dga-f07db0a8-de52-48c4-a115-9d8a4eca379f due to bad geojson - trying these again with spatial.geoJson excluded
[INFO] [08/23/2018 07:31:05.951] [indexer-akka.actor.default-dispatcher-30] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 999 datasets
[INFO] [08/23/2018 07:31:11.543] [indexer-akka.actor.default-dispatcher-23] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:31:15.677] [indexer-akka.actor.default-dispatcher-23] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:31:19.584] [indexer-akka.actor.default-dispatcher-22] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:31:23.696] [indexer-akka.actor.default-dispatcher-32] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:31:30.272] [indexer-akka.actor.default-dispatcher-31] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:31:35.286] [indexer-akka.actor.default-dispatcher-33] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:31:39.789] [indexer-akka.actor.default-dispatcher-33] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:31:45.148] [indexer-akka.actor.default-dispatcher-34] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:31:50.213] [indexer-akka.actor.default-dispatcher-23] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:31:55.496] [indexer-akka.actor.default-dispatcher-23] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:32:00.883] [indexer-akka.actor.default-dispatcher-27] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:32:06.324] [indexer-akka.actor.default-dispatcher-27] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:32:11.643] [indexer-akka.actor.default-dispatcher-23] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:32:15.995] [indexer-akka.actor.default-dispatcher-23] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:32:20.921] [indexer-akka.actor.default-dispatcher-27] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:32:26.195] [indexer-akka.actor.default-dispatcher-36] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:32:30.956] [indexer-akka.actor.default-dispatcher-27] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:32:35.426] [indexer-akka.actor.default-dispatcher-27] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:32:39.786] [indexer-akka.actor.default-dispatcher-27] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:32:43.891] [indexer-akka.actor.default-dispatcher-27] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:32:48.866] [indexer-akka.actor.default-dispatcher-27] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:32:55.219] [indexer-akka.actor.default-dispatcher-35] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:32:59.162] [indexer-akka.actor.default-dispatcher-35] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:33:04.994] [indexer-akka.actor.default-dispatcher-23] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:33:11.094] [indexer-akka.actor.default-dispatcher-35] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:33:15.493] [indexer-akka.actor.default-dispatcher-35] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:33:18.343] [indexer-akka.actor.default-dispatcher-35] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:33:20.662] [indexer-akka.actor.default-dispatcher-35] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:33:23.396] [indexer-akka.actor.default-dispatcher-35] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:33:26.169] [indexer-akka.actor.default-dispatcher-35] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
[INFO] [08/23/2018 07:33:28.932] [indexer-akka.actor.default-dispatcher-35] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1000 datasets
```

### PR DESCRIPTION
**Pull request by mwu2018**
### What this PR does

Fixes #1628

The original design might cause out-of-memory error if the number of datasets is larger than what the memory can hold. The back-pressure was not functioning as expected too. To solve the problems, I designed a new stream source. The source can be filled with datasets even after the stream is running. The number of datasets fetched from the database depends on the indexing speed. Therefore the resource usage is very minimal.

There are some remaining jobs that could be done separately. They are marked as "TODO"s.

### Checklist

-   [x] There are unit tests to verify my changes are correct or unit tests aren't applicable
-   [x] I've updated CHANGES.md with what I changed.
-   [x] I've linked this PR to an issue in ZenHub (core dev team only)

### CODE REVIEW

**Comment by mwu2018** [comment]
I created the branch with wrong name. Branch issue-1682a should have been named as issue-1628a.

**Comment by mwu2018** [comment]
The trimming index delayed 1 second approach seems very appropriate. However it will cause CrawlerApiSpec integration test to fail. An alternative approach, less desirable than the first one, is to use proceedOnConflicts directive in trim function. My experiment shows that it will produce correct results.

**Comment by mwu2018** [comment]
The Travis CI test may take more than 10 mins, resulting timeout failure. I have to manually trigger a rebuild. It took about 13 mins. We need to increase the timeout limit.

File: magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamController.scala [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    * @param nextSize the max size of the batch
+    * @return future optional token. If nonempty, indicate more datasets are available.
+    */
+  def next(nextSize: Int): Future[Option[String]] = {
+    tokenOptionF.flatMap(tokenOption => {
+      if (tokenOption.isEmpty){
+        Future.successful(None)
+      }
+      else {
+        val nextPageF = () => interface.getDataSetsToken(tokenOption.get, bufferSize/2)
```
Comment (t83714) [inline magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamController.scala]: Tested locally via minikube cluster.
Seems indexer will stuck on this line above and terminal printed the following error:
```
[ERROR] [12/03/2018 13:42:42.360] [indexer-akka.actor.default-dispatcher-34] [akka://indexer/user/StreamSupervisor-0/flow-203-0-unknown-operation] Error during preStart in [akka.http.scaladsl.model.HttpEntity$Limitable@7eb90a75]
indexer EntityStreamSizeException: actual entity size (Some(10015105)) exceeded content length limit (8388608 bytes)! You can configure this by setting `akka.http.[server|client].parsing.max-content-length` or calling `HttpEntity.withSizeLimit` before materializing the dataBytes stream.
indexer         at akka.http.scaladsl.model.HttpEntity$Limitable$$anon$1.preStart(HttpEntity.scala:607)
indexer         at akka.stream.impl.fusing.GraphInterpreter.init(GraphInterpreter.scala:520)
indexer         at akka.stream.impl.fusing.GraphInterpreterShell.init(ActorGraphInterpreter.scala:380)
indexer         at akka.stream.impl.fusing.ActorGraphInterpreter.tryInit(ActorGraphInterpreter.scala:538)
indexer         at akka.stream.impl.fusing.ActorGraphInterpreter.preStart(ActorGraphInterpreter.scala:586)
indexer         at akka.actor.Actor$class.aroundPreStart(Actor.scala:510)
indexer         at akka.stream.impl.fusing.ActorGraphInterpreter.aroundPreStart(ActorGraphInterpreter.scala:529)
indexer         at akka.actor.ActorCell.create(ActorCell.scala:590)
indexer         at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:461)
indexer         at akka.actor.ActorCell.systemInvoke(ActorCell.scala:483)
indexer         at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:282)
indexer         at akka.dispatch.Mailbox.run(Mailbox.scala:223)
indexer         at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
indexer         at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
indexer         at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
indexer         at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
indexer         at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```
It looks it tries to request registry for 500 records in one single HTTP request --- It's probably too big...
I think no larger than 50 records in a single HTTP request would be reasonable (before was hard coded to 50 I think)

Comment (mwu2018) [reply magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamController.scala]: The size is configurable in appliation.conf. We can reduce it.

File: magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamSourceController.scala [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    * @param dataSets a batch of datasets to fill the source.
+    * @param hasNext indicating if there are more datasets available
+    * @param isFirst indicating if this is the first batch of datasets
+    */
+  def fillSource(dataSets: Seq[DataSet], hasNext: Boolean, isFirst: Boolean): Unit = {
+    if (isFirst){
+      dataSets foreach (dataSet => {
+        dataSet.copy(publisher = dataSet.publisher)
+        ref ! dataSet
+        if (dataSetCount.incrementAndGet() == bufferSize / 2 && hasNext)
```
Comment (t83714) [inline magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamSourceController.scala]: The logic (that determines the timing of sending `ref ! GET_MORE_DATASETS` message) probably should be moved outside the foreach loop. The condition probably should not be `== bufferSize / 2` --- currently, if indexer fetches less than `bufferSize / 2` records in one request (indexer should fetch much smaller number of dataset in a single request max. 50 as previously comment), the crawling process will hang there.

Comment (mwu2018) [reply magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamSourceController.scala]: Have you verified it?

File: magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamController.scala (hunk 1 above)
Comment (t83714) [reply magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamController.scala]: The problem is that if we change `streamControllerSourceBufferSize` = 100 then crawler will never prepare more than 100 records while elasticsearch processes bulk index request. In the meantime, one elasticsearch bulk request normally consumes 500 ~ 1000 records.

**Comment by AlexGilleran** [comment]
Thanks for reviewing @t83714 !

@mwu2018 once this is ready to be merged would you mind doing a `git rebase -i` to squash these WIP commits?

image

Thanks!

**Comment by t83714** [comment]
@AlexGilleran If you got a chance, could you please have a look at this PR as well? 
As the original author of the crawler, I guess you must be more familiar with the crawling process than me 😄

File: magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamController.scala (hunk 1 above)
Comment (mwu2018) [reply magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamController.scala]: I successfully ran a local test with setting streamControllerSourceBufferSize = 100. 
This parameter is independent of elasticsearch bulk indexing size.

File: magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamSourceController.scala (hunk 2 above)
Comment (t83714) [reply magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamSourceController.scala]: Yes. You are right --- I missed `dataSetCount.incrementAndGet() % (bufferSize / 2) == 0` part and was wondering where we decrease the `dataSetCount`.
Had a few more tests last night and logic here doesn't seems to create any troubles. 👍 
I've also reduced bufferSize to a lower value (100) and the new logic does stops crawling more while elasticsearch was still processing:
```
[INFO] [12/04/2018 09:55:15.824] [indexer-akka.actor.default-dispatcher-10] [RegistryCrawler(akka://indexer)] Crawling registry at /v0
[INFO] [12/04/2018 09:55:22.074] [indexer-akka.actor.default-dispatcher-10] [StreamController(akka://indexer)] Total crawled 1000 datasets from registry
[INFO] [12/04/2018 09:55:24.301] [indexer-akka.actor.default-dispatcher-16] [StreamController(akka://indexer)] Total crawled 1500 datasets from registry
[INFO] [12/04/2018 09:55:24.487] [indexer-akka.actor.default-dispatcher-14] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1 datasets
[INFO] [12/04/2018 09:55:55.156] [indexer-akka.actor.default-dispatcher-2] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 500 datasets
[INFO] [12/04/2018 09:55:56.711] [indexer-akka.actor.default-dispatcher-2] [StreamController(akka://indexer)] Total crawled 2000 datasets from registry
[INFO] [12/04/2018 09:56:17.545] [indexer-akka.actor.default-dispatcher-17] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 500 datasets
[INFO] [12/04/2018 09:56:18.978] [indexer-akka.actor.default-dispatcher-2] [StreamController(akka://indexer)] Total crawled 2500 datasets from registry
[INFO] [12/04/2018 09:56:51.148] [indexer-akka.actor.default-dispatcher-10] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 500 datasets
[INFO] [12/04/2018 09:56:52.707] [indexer-akka.actor.default-dispatcher-8] [StreamController(akka://indexer)] Total crawled 3000 datasets from registry
[INFO] [12/04/2018 09:57:22.418] [indexer-akka.actor.default-dispatcher-2] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 500 datasets
[INFO] [12/04/2018 09:57:24.139] [indexer-akka.actor.default-dispatcher-15] [StreamController(akka://indexer)] Total crawled 3500 datasets from registry
[INFO] [12/04/2018 09:57:38.986] [indexer-akka.actor.default-dispatcher-32] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 500 datasets
[INFO] [12/04/2018 09:57:40.657] [indexer-akka.actor.default-dispatcher-15] [StreamController(akka://indexer)] Total crawled 4000 datasets from registry
```
Well done & great job~ 🎉

File: magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/search/elasticsearch/ElasticSearchIndexer.scala [hunk 3]
```diff
@@ -375,27 +369,9 @@ class ElasticSearchIndexer(
...
-            thisResultFuture.map { _ =>
-              combinedResult.copy(successes = combinedResult.successes + 1)
-            }.recover {
-              case (e: Throwable) =>
-                combinedResult.copy(failures = combinedResult.failures :+ thisResultIdentifier)
-            }
-          }
-      })
+  // TODO: Instead of using indexQueue that performs bulk indexing, make single indexing request.
+  private def index2(dataSet: DataSet, promise: Promise[Unit] = Promise[Unit]): Future[Unit] = {
```
Comment (AlexGilleran) [inline magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/search/elasticsearch/ElasticSearchIndexer.scala]: Can we name this something more descriptive? What does it do that's different from `index` ?

File: magda-indexer/src/test/scala/au/csiro/data61/magda/indexer/helpers/StreamControllerTest.scala [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    val numOfDataSets = 10 * bufferSize - 1
+    run(numOfDataSets)
+  }
+
+  "The stream controller" should "support the indexer stream 2" in {
+    val numOfDataSets = 10 * bufferSize
+    run(numOfDataSets)
+  }
+
+  "The stream controller" should "support the indexer stream 3" in {
```
Comment (AlexGilleran) [inline magda-indexer/src/test/scala/au/csiro/data61/magda/indexer/helpers/StreamControllerTest.scala]: Can we get some more descriptive test descriptions - what is this test doing that the other ones aren't? If this test is the only one that fails I'm going to have no idea what it specifically does that's failing.

File: magda-indexer/src/test/scala/au/csiro/data61/magda/indexer/helpers/StreamSourceControllerTest.scala [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+  override def afterEach(): Unit = {
+    super.afterEach()
+    actorRef ! PoisonPill
+  }
+
+  "The stream source controller" should "fill the source after the stream starts" in {
+    val actualDataSetsF: Future[Seq[DataSet]] = source.runWith(Sink.seq)
+    ssc.fillSource(dataSets, false, true)
+//    ssc.terminate()
```
Comment (AlexGilleran) [inline magda-indexer/src/test/scala/au/csiro/data61/magda/indexer/helpers/StreamSourceControllerTest.scala]: Do we still need these commented lines?

File: magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamSourceController.scala [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    * Fill the source with given datasets
+    *
+    * @param dataSets a batch of datasets to fill the source.
+    * @param hasNext indicating if there are more datasets available
+    * @param isFirst indicating if this is the first batch of datasets
+    */
+  def fillSource(dataSets: Seq[DataSet], hasNext: Boolean, isFirst: Boolean): Unit = {
+    if (isFirst){
+      dataSets foreach (dataSet => {
+        dataSet.copy(publisher = dataSet.publisher)
```
[the remark this answers is not in the document]
Comment (mwu2018) [reply magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamSourceController.scala]: This is from the old implementation and I was puzzled when I copied it. Probably not needed.

File: magda-indexer/src/test/scala/au/csiro/data61/magda/indexer/helpers/StreamSourceControllerTest.scala (hunk 5 above)
Comment (mwu2018) [reply magda-indexer/src/test/scala/au/csiro/data61/magda/indexer/helpers/StreamSourceControllerTest.scala]: No, I will remove them.

File: magda-indexer/src/test/scala/au/csiro/data61/magda/indexer/helpers/StreamControllerTest.scala (hunk 4 above)
Comment (mwu2018) [reply magda-indexer/src/test/scala/au/csiro/data61/magda/indexer/helpers/StreamControllerTest.scala]: They are testing the control logic at edge cases. Given a source buffer size, it should work with any number of datasets. It is easy to make logical errors for those dataset numbers. How about changing them like this:
"The stream controller" should "support the indexer stream when the number of datasets is xxx".

File: magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/search/elasticsearch/ElasticSearchIndexer.scala (hunk 3 above)
Comment (mwu2018) [reply magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/search/elasticsearch/ElasticSearchIndexer.scala]: The original implementation uses the name "index", the same as another "index" that is api. I only renamed it to "index2". I can revert it back if you prefer. As described in the TODO comment, this part will be re-written. I have not figured out how to do it yet.

File: magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamSourceController.scala [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+                             implicit val config: Config,
+                             implicit val materializer: Materializer) {
+
+  val log = Logging(system, getClass)
+  implicit val ec: ExecutionContextExecutor = system.dispatcher
+  implicit val scheduler: Scheduler = system.scheduler
+
+  private val GET_MORE_DATASETS: String = "Get more datasets"
+  private val NO_MORE_DATASETS: String = "No more datasets"
+  private val dataSetCount = new AtomicLong(0)
```
[the remark this answers is not in the document]
Comment (mwu2018) [reply magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamSourceController.scala]: ok

File: magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamSourceController.scala [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+          }
+          else
+            push(out, message)
+        }
+
+        override def onPull(): Unit = {
+          pull(in)
+        }
+
+       setHandlers(in, out, this)
```
[the remark this answers is not in the document]
Comment (mwu2018) [reply magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamSourceController.scala]: You have eagle eyes 👍

File: magda-indexer/src/test/scala/au/csiro/data61/magda/indexer/helpers/StreamControllerTest.scala (hunk 4 above)
Comment (AlexGilleran) [reply magda-indexer/src/test/scala/au/csiro/data61/magda/indexer/helpers/StreamControllerTest.scala]: 👍

Comment (mwu2018) [reply magda-indexer/src/test/scala/au/csiro/data61/magda/indexer/helpers/StreamControllerTest.scala]: Please note that StreamControllerTest was written in the early stage when MockIndexer still tries to mock the old index function. However, the control logic still holds.

File: magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamController.scala (hunk 1 above)
Comment (t83714) [reply magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamController.scala]: For this one, we will be fine as long we change `streamControllerSourceBufferSize` to 100:
```
crawler {
  streamControllerSourceBufferSize = 100
}
```
